### PR TITLE
feat: new client

### PR DESCRIPTION
**Pull request by AlexanderN8**

### CODE REVIEW

**Review by allacherkes — CHANGES_REQUESTED** [review]

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@sap-devx/feature-toggle-node",
-  "version": "1.0.11",
+  "version": "1.0.12",
```
[the remark this answers is not in the document]
Comment (AlexanderN8) [reply package.json]: Yes, It's done automatically by npm. I will revert it.

File: package.json [hunk 2]
```diff
@@ -30,22 +30,22 @@
   "license": "Apache-2.0",
   "dependencies": {
     "await-to-js": "3.0.0",
-    "parse-duration": "0.4.4",
-    "unleash-client": "3.4.0"
+    "node-cache": "^5.1.2",
```
[the remark this answers is not in the document]
Comment (AlexanderN8) [reply package.json]: done

File: src/api.ts [hunk 3]
```diff
@@ -1,35 +1,80 @@
...
-import { Unleash } from "unleash-client";
+import { requestFeatureToggles } from "./request";
 
-export async function isFeatureEnabled(extensionName: string, featureToggleName: string): Promise<boolean> {
-  log(`Checking if Extension Name: "${extensionName}", Feature Toggle Name: "${featureToggleName}" is enabled`);
+import { isToggleEnabled } from "./strategy_matching";
+import { Cache } from "./cache";
 
-  const ftName = `${extensionName}.${featureToggleName}`;
+export interface Parameters {
```
[the remark this answers is not in the document]
Comment (AlexanderN8) [reply src/api.ts]: done

File: src/api.ts [hunk 4]
```diff
@@ -1,35 +1,80 @@
...
 
-    //get unleash client
-    const client: Unleash = await clientManager.getUnleashClient(extensionName);
+    const toggles: Features = await getFeatureToggles();
+    if (toggles.features.length) {
+      Cache.setFeatureToggles(toggles);
 
-    // get the context
-    const context: Context = contextManager.getContext(extensionName);
+      const toggle: Toggle | undefined = findToggleByName(toggles, ftName);
```
[the remark this answers is not in the document]
Comment (AlexanderN8) [reply src/api.ts]: comment just repeat function name. It's enough that the function contains a lot of explanations

File: src/cache.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import * as NodeCache from "node-cache";
+import { Features } from "./api";
+
+const REFRESH_INTERVAL = 60 * 15; //15 minutes
```
[the remark this answers is not in the document]
Comment (AlexanderN8) [reply src/cache.ts]: yes. Things that have an impact on the workload of the server should be defined by code.

File: src/request.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        res.on("data", (chunk) => {
+          data.push(chunk);
+        });
+
+        res.on("end", () => {
+          resolve(JSON.parse(Buffer.concat(data).toString()));
+        });
+      })
+      .on("error", (e) => {
+        console.error(e);
```
[the remark this answers is not in the document]
Comment (AlexanderN8) [reply src/request.ts]: done

File: src/request.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+const SERVER_ENDPOINT = "http://ft-manager.feature-toggle.svc.cluster.local/v1/features";
+
+export function requestFeatureToggles(): Promise<Features> {
+  return new Promise((resolve, reject) => {
+    https
+      .get(SERVER_ENDPOINT, (res) => {
+        log("Get Toggles from server with Status Code: " + res.statusCode);
+
+        const data: Uint8Array[] = [];
```
[the remark this answers is not in the document]
Comment (AlexanderN8) [reply src/request.ts:20]: take response from server and change to json

File: src/request.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import { log } from "./logger";
+import * as https from "http";
+import { Features } from "./api";
+
+const SERVER_ENDPOINT = "http://ft-manager.feature-toggle.svc.cluster.local/v1/features";
```
[the remark this answers is not in the document]
Comment (AlexanderN8) [reply src/request.ts]: done

File: src/cache.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+const FEATURES_KEY = "features";
+const ftCache = new NodeCache({ stdTTL: REFRESH_INTERVAL });
+
+export class Cache {
```
[the remark this answers is not in the document]
Comment (AlexanderN8) [reply src/cache.ts:13]: done

File: src/strategy_matching.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  landscapes = "landscapes",
+  subaccounts = "subaccounts",
+  users = "users",
+  wss = "wss",
+  tenantids = "tenantids",
+}
+
+// TODO: Check infrastructure
+function isMatchStrategies(toggle: Toggle): boolean {
+  const currentContext: IterateableContext = createContextEntity();
```
[the remark this answers is not in the document]
Comment (AlexanderN8) [reply src/strategy_matching.ts]: it's a small effort for the system, but it won't be possible to update env variables after the client started.

File: src/strategy_matching.ts [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+  for (const parameterName of Object.values(ParametersNames)) {
+    const parameterValue = toggle[parameterName];
+
+    const singularParamName = convertPluralNameToSingular(parameterName);
+    const currentContextParamValue = currentContext[singularParamName];
+
+    if (parameterValue?.length && !parameterValue?.includes(currentContextParamValue)) {
```
[the remark this answers is not in the document]
Comment (AlexanderN8) [reply src/strategy_matching.ts]: changed

File: src/api.ts [hunk 12]
```diff
@@ -1,35 +1,80 @@
...
-      throw new Error("Feature toggle name can not be empty, null or undefined");
+    if (toggleFromCache != undefined) {
+      return toggleFromCache;
     }
 
-    //get unleash client
-    const client: Unleash = await clientManager.getUnleashClient(extensionName);
+    const toggles: Features = await getFeatureToggles();
+    if (toggles.features.length) {
+      Cache.setFeatureToggles(toggles);
```
[the remark this answers is not in the document]
Comment (AlexanderN8) [reply src/api.ts]: removed

**Review by allacherkes — COMMENTED** [review]
Please update readme as well
Reviewed only code part.

File: src/api.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (allacherkes) [reply src/api.ts]: I meant comments inside function

File: test/api.spec.ts [hunk 13]
```diff
@@ -1,20 +1,15 @@
 import { expect } from "chai";
```
Comment (allacherkes) [inline test/api.spec.ts:1]: Is it possible to do tests for positive flow? It was there before

File: test/client.spec.ts [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+  it("toggle name found in cache by key - return true", async () => {
+    const ftName = "ext.ftName";
+
+    sinon.stub(Cache, "getToggleByKey").returns(true);
+
+    const isEnabled = await Client.findToggleAndReturnState(ftName);
+    expect(isEnabled).to.be.true;
+  });
+
+  function stubDependencies(features: Client.Features): void {
```
Comment (allacherkes) [inline test/client.spec.ts:23]: move this function up

File: test/client.spec.ts [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+  });
+
+  it("response return empty feature toggles list - return false", async () => {
+    const ftName = "ext.ftName";
+    const features: Client.Features = {
+      features: [],
+    };
+
+    sinon.stub(Request, "requestFeatureToggles").resolves(features);
+    sinon.stub(Strategies, "isToggleEnabled").returns(true);
```
Comment (allacherkes) [inline test/client.spec.ts:75]: some of stubs are not required here

File: test/client.spec.ts [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+    sinon.stub(Cache, "getFeatureToggles").returns(undefined);
+    sinon.stub(Cache, "getToggleByKey").returns(undefined);
+    sinon.stub(Cache, "setTogglesByKey").returns();
+    const isEnabled = await Client.findToggleAndReturnState(ftName);
+    expect(isEnabled).to.be.false;
+  });
+
+  it("gets feature toggles list from cache - return false", async () => {
```
Comment (allacherkes) [inline test/client.spec.ts:83]: description is not correst

File: test/current_context.spec.ts [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+      LANDSCAPE_ENVIRONMENT: "test_landscape",
+      LANDSCAPE_NAME: "test_landscape_name",
+      TENANT_ID: "test_tenant_id",
+      TENANT_NAME: "test_tenant_name",
+      WORKSPACE_ID: "test_worksscape_id",
+    };
+    sinon.stub(process, "env").value(envValues);
+
+    const currentContext = createContextEntity();
+    logger.log("some message");
```
Comment (allacherkes) [inline test/current_context.spec.ts:24]: what for?

File: test/current_context.spec.ts [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+import * as logger from "../src/logger";
+import { expect } from "chai";
+import { createContextEntity } from "../lib/current_context";
+
+describe("createContextEntity", () => {
+  afterEach(() => {
+    sinon.restore();
+  });
+
+  it("save env variables and check that equal", () => {
```
Comment (allacherkes) [inline test/current_context.spec.ts:12]: infrastracture is missing

File: test/current_context.spec.ts [hunk 19]
```diff
@@ -0,0 +1,32 @@
+import { afterEach, describe, it } from "mocha";
+import * as sinon from "sinon";
+import * as logger from "../src/logger";
+import { expect } from "chai";
+import { createContextEntity } from "../lib/current_context";
```
Comment (allacherkes) [inline test/current_context.spec.ts:5]: I get error here

File: test/strategies.spec.ts [hunk 20]
```diff
@@ -0,0 +1,177 @@
+import { afterEach, describe, it } from "mocha";
+import * as sinon from "sinon";
+import * as CurrentContext from "../src/current_context";
+import { expect } from "chai";
+import { IterateableContext, isToggleEnabled } from "../src/strategies";
+import { Toggle } from "../src/client";
+import { ContextData } from "../lib/current_context";
```
Comment (allacherkes) [inline test/strategies.spec.ts:7]: error here

File: test/strategies.spec.ts [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+      tenantid: "",
+    } as ContextData;
+
+    sinon.stub(CurrentContext, "createContextEntity").returns(contextParams as ContextData);
+
+    const isEnabled = await isToggleEnabled(toggle);
+    expect(isEnabled).to.be.false;
+  });
+
+  it("toggle enabled and has one correct strategies except tenantid return false", async () => {
```
Comment (allacherkes) [inline test/strategies.spec.ts:118]: not correct description

**Review by allacherkes — CHANGES_REQUESTED** [review]
tests code review

File: test/current_context.spec.ts (hunk 19 above)
Comment (AlexanderN8) [reply test/current_context.spec.ts:5]: Please describe error

File: test/strategies.spec.ts (hunk 20 above)
Comment (AlexanderN8) [reply test/strategies.spec.ts:7]: Please describe error

File: test/current_context.spec.ts (hunk 19 above)
Comment (allacherkes) [reply test/current_context.spec.ts]: cant find the  path. (pat is not correct)

File: test/strategies.spec.ts (hunk 20 above)
Comment (allacherkes) [reply test/strategies.spec.ts:7]: cant find the path. (pat is not correct)

File: test/api.spec.ts (hunk 13 above)
Comment (AlexanderN8) [reply test/api.spec.ts:1]: done

File: test/client.spec.ts (hunk 14 above)
Comment (AlexanderN8) [reply test/client.spec.ts:19]: done

File: test/client.spec.ts (hunk 15 above)
Comment (AlexanderN8) [reply test/client.spec.ts:81]: done

File: test/client.spec.ts (hunk 16 above)
Comment (AlexanderN8) [reply test/client.spec.ts]: done

File: test/current_context.spec.ts (hunk 17 above)
Comment (AlexanderN8) [reply test/current_context.spec.ts]: done

File: test/current_context.spec.ts (hunk 18 above)
Comment (AlexanderN8) [reply test/current_context.spec.ts:11]: added

File: test/current_context.spec.ts (hunk 19 above)
Comment (AlexanderN8) [reply test/current_context.spec.ts]: updated

File: test/strategies.spec.ts (hunk 20 above)
Comment (AlexanderN8) [reply test/strategies.spec.ts:7]: updated

File: test/strategies.spec.ts (hunk 21 above)
Comment (AlexanderN8) [reply test/strategies.spec.ts]: updated

File: src/api.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (AlexanderN8) [reply src/api.ts]: added

File: test/current_context.spec.ts (hunk 19 above)
Comment (allacherkes) [reply test/current_context.spec.ts]: not yet :)

File: test/strategies.spec.ts (hunk 20 above)
Comment (allacherkes) [reply test/strategies.spec.ts:7]: not yet :)

File: README.md [hunk 22]
```diff
@@ -38,20 +38,19 @@ Install the feature-toggle-node as a dependency.
...
-- `FT_SERVER_ENDPOINT` - Feature toggle server endpoint (mandatory)
+- `FTM_HOST` - Feature toggle server endpoint (mandatory)
 - `USER_NAME` - Name of the user logged into SAP Business Application Studio (mandatory)
 - `WORKSPACE_ID` - ID of the workspace (mandatory)
 - `TENANT_ID` - (mandatory)
 - `TENANT_NAME` - (mandatory)
 - `LANDSCAPE_ENVIRONMENT` - (mandatory)
 - `LANDSCAPE_NAME` - (mandatory)
-- `FT_TOKEN` - Value of the feature toggle server API token (optional)
 - `SHOW_LOG` - If true, displays console logs (optional)
```
Comment (allacherkes) [inline README.md:48]: What about LANDSCAPE_INFRASTRUCTURE

**Review by allacherkes — APPROVED** [review]

**Pull request merged**